content/4_Gold/PURS.mdx

```diff
@@ -95,6 +92,105 @@ using std::endl;
 using std::vector;
 
 /** A data structure that can answer point update & range minimum queries. */
+template <class T> class MinSegmentTree {
+  private:
+	const T DEFAULT = std::numeric_limits<T>().max();
+
+	int len;
+	int padded_len;
+	vector<T> segtree;
+
+	T combine(const T &a, const T &b) { return std::min(a, b); }
+
+	void set(int ind, T val, int at, int at_left, int at_right) {
+		if (at_right - at_left == 1) {
+			segtree[at] = val;
+			return;
+		}
+		int mid = (at_left + at_right) / 2;
+		if (ind < mid) {
+			set(ind, val, 2 * at + 1, at_left, mid);
+		} else {
+			set(ind, val, 2 * at + 2, mid, at_right);
+		}
+		segtree[at] = combine(segtree[2 * at + 1], segtree[2 * at + 2]);
+	}
+
+	T range_min(int start, int end, int at, int at_left, int at_right) {
+		if (at_right <= start || end <= at_left) { return DEFAULT; }
+		if (start <= at_left && at_right <= end) { return segtree[at]; }
+		int mid = (at_left + at_right) / 2;
+		T left_res = range_min(start, end, 2 * at + 1, at_left, mid);
+		T right_res = range_min(start, end, 2 * at + 2, mid, at_right);
+		return combine(left_res, right_res);
+	}
+
+  public:
+	MinSegmentTree(int len) : len(len) {
+		padded_len = 1;
+		while (padded_len < len) { padded_len *= 2; }
```

Comment:
you could probably use `__lg` or `std::bit_width` (if C++ 20)

Reply:
given that i've only heard of this function today,,, not sure if i should use it
also usaco doesn't have c++ 20 yet

Reply:
yeah i probably would not use std::bit_width

if ur fine w/ using `__lg` (I think we use some of the GCC defined __ functions like `__pop_count` in our modules, so maybe this is ok), then I think its good